### ISSUE TITLE
Transfer token via cheque (ERC20 Extension)

### ISSUE THREAD

**Issue by chompomonim** [ethereum/EIPs#964]
The problem of token spending is that token owner have to have eth to pay gas. So it's impossible to have pure tokens account.

There are situations when service is ready to cover gas expenses while preserving user's ability to spend her tokens at arbitrary moment. i.e. initiate logical transaction.

There could be some kind of `transferByCheque` call, where anyone who has token owner's signed cheque (v, r, s) can move tokens.

Possible functions realisation:
```solidity
    function transferByCheque(address _to, uint256 _value, uint256 _nonce, uint8 v, bytes32 r, bytes32 s) public returns (bool) {
        require(_to != address(this));

        // Check if signature is valid, get signer's address and mark this cheque as used.
        bytes memory prefix = "\x19Ethereum Signed Message:\n32";
        bytes32 hash = keccak256(prefix, keccak256(_to, _value, _nonce));

        address signer = ecrecover(hash, v, r, s);
        require (signer != 0);
        require (!usedNonces[signer][_nonce]);
        usedNonces[signer][_nonce] = true;

        // Transfer tokens
        balanceOf[signer] = balanceOf[signer].sub(_value);
        balanceOf[_to] = balanceOf[_to].add(_value);
        Transfer(signer, _to, _value, _data);

        return true;
    }
```

On user's (wallet) side cheque creation could look like:
```js
const leftPad = require('left-pad')

const hexData = [
     _to.slice(2),
     leftPad((_value).toString(16), 64, 0),
     leftPad((_nonce).toString(16), 64, 0)
].join('')

const msg = web3.sha3(hexData, { encoding: 'hex' }).slice(2)

const signature = web3.eth.sign(accounts[0], msg).slice(2)

const r = '0x' + signature.slice(0,64);
const s = '0x' + signature.slice(64, 128);
const v = Number.parseInt(signature.slice(128, 130), 16) + 27;
```

Later transaction could look like:
```js
await token.transferByCheque(_to, _value, _nonce, v, r, s)
```

I think this kind of cheques could potentially be widely used. Example use-case:

Imagine situation where shop gives discount points for client in a form of tokens. Also client downloads special app which is not only loyalty app but also is some kind of wallet and stores private key.

Later, when client will want to use such points, without depositing some amount of gwei into his 'token wallet', he will not be able to transfer tokens. Meanwhile this token wallet could sign cheque and transfer it to shop back. Then shop (or some another entity) using this cheque could transfer tokens while paying for gas by himself.

**Comment by ksystemsio** [comment]
Please re-open

**Comment by chompomonim** [comment]
@ksystemsio this task was closed, because it is very similar to #965 and I do believe that any advanced extensions should be done for #777 as it is better standard for tokens (IMHO).

Could we have any discussions on #965 first and if we will agree that it's needed for ERC20, then we can reopen this task.
